rasterio/__init__.py

```diff
@@ -202,38 +206,69 @@ def open(path, mode='r', driver=None, width=None, height=None,
             "Blacklisted: file cannot be opened by "
             "driver '{0}' in '{1}' mode".format(driver, mode))
 
-    _, _, scheme = parse_path(path)
-
-    with Env() as env:
-        # Get AWS credentials only if we're attempting to access a
-        # raster using the S3 scheme.
-        if scheme == 's3':
-            env.get_aws_credentials()
-            log.debug("AWS credentials have been obtained")
-
-        # Create dataset instances and pass the given env, which will
-        # be taken over by the dataset's context manager if it is not
-        # None.
-        if mode == 'r':
-            s = DatasetReader(path)
-        elif mode == 'r-':
-            warnings.warn("'r-' mode is deprecated, use 'r'",
-                          DeprecationWarning)
-            s = DatasetReader(path)
-        elif mode == 'r+':
-            s = get_writer_for_path(path)(path, mode)
-        elif mode == 'w':
-            s = get_writer_for_driver(driver)(path, mode, driver=driver,
-                                              width=width, height=height,
-                                              count=count, crs=crs,
-                                              transform=transform,
-                                              dtype=dtype, nodata=nodata,
-                                              **kwargs)
-        else:
-            raise ValueError(
-                "mode string must be one of 'r', 'r+', or 'w', not %s" % mode)
-        s.start()
-        return s
+    # Special case for file object argument.
+    if mode =='r' and hasattr(fp, 'read'):
+
+        @contextmanager
+        def fp_reader(fp):
+            memfile = MemoryFile(fp.read())
+            dataset = memfile.open()
+            yield dataset
```

Comment:
`@contextmanager` is new to me as well. Yielding makes for much nicer syntax than defining a class with `__enter__` and `__exit__` methods.